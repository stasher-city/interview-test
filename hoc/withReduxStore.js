import PropTypes from 'prop-types'
import React, { Component } from 'react'

import { initializeStore } from '../redux/index'


const isServer = typeof window === 'undefined'
const __NEXT_REDUX_STORE__ = '__NEXT_REDUX_STORE__'


export function getOrCreateStore(initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[__NEXT_REDUX_STORE__]) {
    window[__NEXT_REDUX_STORE__] = initializeStore(initialState)
  }

  return window[__NEXT_REDUX_STORE__]
}


export default (TopApp) => {
  class AppWithRedux extends Component {
    static propTypes = { initialReduxState: PropTypes.object };

    static async getInitialProps (appCtx) {
      const store = getOrCreateStore()

      appCtx.ctx.store = store

      let appProps = {}

      if (typeof TopApp.getInitialProps === 'function') {
        appProps = await TopApp.getInitialProps(appCtx)
      }

      return {
        ...appProps,
        initialReduxState: store.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return (
        <TopApp {...this.props} reduxStore={this.reduxStore} />
      )
    }
  }

  return AppWithRedux
}
