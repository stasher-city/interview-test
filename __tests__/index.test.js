import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'

import Home from '../pages/index'

import { getOrCreateStore } from '../hoc/withReduxStore'

const store = getOrCreateStore()
const component = mount(
  <Provider store={store}>
    <Home />
  </Provider>
)


describe('Index', () => {
  it('renders properly', () => {
    expect.assertions(1)

    expect(component.html()).toMatchSnapshot()
  })
})
