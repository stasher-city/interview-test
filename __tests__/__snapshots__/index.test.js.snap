// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index renders properly 1`] = `"<div class=\\"page\\"><h3 class=\\"title\\">Stasher!</h3><header class=\\"box\\"><input class=\\"item\\" type=\\"search\\" value=\\"london\\"><button class=\\"item\\">Search</button><hr class=\\"hr\\"><div class=\\"filtersWrapper\\"><div class=\\"filterButton item\\"><label class=\\"label\\">Bags</label><br><select class=\\"selector\\"><option value=\\"0\\">1</option><option value=\\"1\\">2</option><option value=\\"2\\">3</option><option value=\\"3\\">4</option><option value=\\"4\\">5</option><option value=\\"5\\">6</option><option value=\\"6\\">7</option><option value=\\"7\\">8</option><option value=\\"8\\">9</option><option value=\\"9\\">10</option><option value=\\"10\\">11</option><option value=\\"11\\">12</option><option value=\\"12\\">13</option><option value=\\"13\\">14</option><option value=\\"14\\">15</option><option value=\\"15\\">16</option><option value=\\"16\\">17</option><option value=\\"17\\">18</option><option value=\\"18\\">19</option><option value=\\"19\\">20</option></select></div><div class=\\"filterButton item\\"><label class=\\"label\\">Drop off</label><br><div><input type=\\"date\\" value=\\"2019-11-15\\"><select class=\\"timeSelector\\" name=\\"Drop off\\"><option value=\\"0\\">00:00</option><option value=\\"1\\">01:00</option><option value=\\"2\\">02:00</option><option value=\\"3\\">03:00</option><option value=\\"4\\">04:00</option><option value=\\"5\\">05:00</option><option value=\\"6\\">06:00</option><option value=\\"7\\">07:00</option><option value=\\"8\\">08:00</option><option value=\\"9\\">09:00</option><option value=\\"10\\">10:00</option><option value=\\"11\\">11:00</option><option value=\\"12\\">12:00</option><option value=\\"13\\" selected=\\"\\">13:00</option><option value=\\"14\\">14:00</option><option value=\\"15\\">15:00</option><option value=\\"16\\">16:00</option><option value=\\"17\\">17:00</option><option value=\\"18\\">18:00</option><option value=\\"19\\">19:00</option><option value=\\"20\\">20:00</option><option value=\\"21\\">21:00</option><option value=\\"22\\">22:00</option><option value=\\"23\\">23:00</option></select></div></div><div class=\\"filterButton item\\"><label class=\\"label\\">Pick up</label><br><div><input type=\\"date\\" value=\\"2019-11-15\\"><select class=\\"timeSelector\\" name=\\"Pick up\\"><option value=\\"0\\">00:00</option><option value=\\"1\\">01:00</option><option value=\\"2\\">02:00</option><option value=\\"3\\">03:00</option><option value=\\"4\\">04:00</option><option value=\\"5\\">05:00</option><option value=\\"6\\">06:00</option><option value=\\"7\\">07:00</option><option value=\\"8\\">08:00</option><option value=\\"9\\">09:00</option><option value=\\"10\\">10:00</option><option value=\\"11\\">11:00</option><option value=\\"12\\">12:00</option><option value=\\"13\\">13:00</option><option value=\\"14\\" selected=\\"\\">14:00</option><option value=\\"15\\">15:00</option><option value=\\"16\\">16:00</option><option value=\\"17\\">17:00</option><option value=\\"18\\">18:00</option><option value=\\"19\\">19:00</option><option value=\\"20\\">20:00</option><option value=\\"21\\">21:00</option><option value=\\"22\\">22:00</option><option value=\\"23\\">23:00</option></select></div></div></div></header><div><ul></ul></div></div>"`;
